### DISCUSSION TITLE
GAP 4.14.0-beta1

### DISCUSSION THREAD

**Discussion by fingolfin** [gap-system/gap#5843] [General]
Dear all,

I plan to release GAP 4.14.0 in 1-2 weeks (before December 1st) assuming all goes well. To prepare for that, I've made a prerelease to allow testing.

Version 4.14.0-beta1 is mostly intended for use by people who package GAP for Linux distributions, or similar things (Homebrew, FreeBSD ports, Conda, SageMath, ...). The idea is to iron out issues related to packaging before the final release.

That said, everyone else is of course also welcome to give it a try. I don't expect major changes in the final release, though we'll still merge PRs and package updates if they don't appear "risky".

You can find this pre-release at https://github.com/gap-system/gap/releases/tag/v4.14.0-beta1

If you want all GAP sources and all packages, you need [gap-4.14.0-beta1.tar.gz](https://github.com/gap-system/gap/releases/download/v4.14.0-beta1/gap-4.14.0-beta1.tar.gz).

Some people who I know package GAP and might be interested in this include @antonio-rojas @isuruf @saraedum @jamesjer @orlitzky @collares @AndreasEnge @yurivict @kiwifb @mkoeppe

**Comment by cdwensley** [comment]
This is just a report of what went wrong for me yesterday in case it is of any help. 
Received the release announcement and downloaded the source and packages separately.
Made the binary and built the packages.  I _think_ GAP then. loaded correctly. 
I then deleted those packages which I have cloned from GitHub, so that my clones would be used instead. 
I then updated these clones (> git pull upstream master).
Running GAP then produced a string or errors referring to SwapMatrixRows/Columns (see below). 
I realise there have been recent PRs dealing with these operations. 
I imagine that one of these packages might be causing the problem, but have no idea which. 
The only cloned packages which appear in the list below are FGA and Polycyclic.  Chris

***@***.*** gap-4.14.0 % ./gap
 ┌───────┐   GAP 4.13.1 of 2024-06-11
 │  GAP  │   https://www.gap-system.org
 └───────┘   Architecture: aarch64-apple-darwin21-default64-kv9
 Configuration:  gmp 6.2.1, GASMAN
 Loading the library Error, Variable: 'SWAP_MAT_ROWS' must have a value
not in any function at /Users/christopherwensley/gap/gap-4.14.0/lib/matobj2.gd:2037
Error, Variable: 'SWAP_MAT_COLS' must have a value
not in any function at /Users/christopherwensley/gap/gap-4.14.0/lib/matobj2.gd:2054
and packages ...
Syntax warning: Unbound global variable in /Users/christopherwensley/gap/gap-4\
.14.0/lib/matrix.gi:1493
                SwapMatrixRows(mat, j, i);
                ^^^^^^^^^^^^^^
…
(14 more error messages deleted)
...
 Packages:   AClib 1.3.2, Alnuth 3.2.1, AtlasRep 2.1.9, AutPGrp 1.11, 
             Browse 1.8.21, CaratInterface 2.3.7, CRISP 1.4.6, Cryst 4.1.27, 
             CrystCat 1.1.10, CTblLib 1.3.9, curlInterface 2.4.0, 
             FactInt 1.6.3, FGA 1.5.0, GAPDoc 1.6.7, IO 4.9.0, IRREDSOL 1.4.4, 
             LAGUNA 3.9.7, Polenta 1.3.10, Polycyclic 2.17dev, PrimGrp 3.4.4, 
             RadiRoot 2.9, ResClasses 4.7.3, SmallGrp 1.5.4, Sophus 1.27, 
             SpinSym 1.5.2, StandardFF 1.0, TomLib 1.2.11, TransGrp 3.6.5, 
             utils 0.87
 Try '??help' for help. See also '?copyright', '?cite' and '?authors'
gap>

**Comment by ChrisJefferson** [comment]
The problem here is (I think) you haven't rebuilt the GAP kernel (try typing `make distclean; ./configure; make`. I can tell this because GAP is reporting `4.13.1`. The issues aren't in packages, as they are coming from `lib/matobj2.gd`.

**Comment by cdwensley** [comment]
Yes, that fixes it.  Don’t understand what happened, but don’t need to I guess.  Thanks.

**Comment by yurivict** [comment]
I verified that 4.14.0-beta1 builds and passes all tests from the FreeBSD port's perspective.

**Reply by fingolfin** [reply]
Thank you!

**Comment by jamesjer** [comment]
I did some test builds for Fedora and everything seems okay so far.  I have not tried building all of the packages yet, but will do so over the next few days.

**Reply by fingolfin** [reply]
Thank you @jamesjer

**Comment by orlitzky** [comment]
We have some fresh Gentoo bugs for GCC-15 compatibility,

* https://bugs.gentoo.org/943925
* https://bugs.gentoo.org/944356
* https://bugs.gentoo.org/944036

due to `-Werror=incompatible-pointer-types` and the switch to C23 as the default standard (which e.g. has its own `bool` type). I've built my own GCC-15 and can confirm that the problems persist in the latest versions of things, including the beta.

If the [development roadmap](https://gcc.gnu.org/develop.html) can be trusted, GCC-15 is in stage three (out of four) of its release cycle, and could be released while GAP 4.14.0 is current (typically April/May). Now might be the time to either fix what it is complaining about, or append the necessary `-Wno-error` and `-std` flags to the builds system.

**Comment by orlitzky** [comment]
With GCC 14.2.1, things are mostly OK.

I somehow have test failures in hap-1.66 even with GAP 4.13.1:

```
testing: /var/lib/portage/tmp/portage/dev-gap/hap-1.66/temp/pkg/hap-1.66/tst/testallV11/poin.tst
   ? `intvec`(GAP_intvec) = `bigintvec` is not supported
   ? expected `intvec` = `intvec`
   ? error occurred in or before STDIN line 6948: `, 1, intvec(1,1,1,2,2,3,4) );`
   ? expected intvec-expression. type 'help intvec;'
########> Diff in /var/lib/portage/tmp/portage/dev-gap/hap-1.66/temp/pkg/hap-1.66/tst/testallV11/poin.tst:3
# Input is:
P:=PoincareSeriesLHS(G);
# Expected output:
(1)/(-x_1^3+3*x_1^2-3*x_1+1)
# But found:
Error, no method found! For debugging hints type ?Recovery from NoMethodFound
Error, no 1st choice method found for `Remove' on 2 arguments
########
########> Diff in /var/lib/portage/tmp/portage/dev-gap/hap-1.66/temp/pkg/hap-1.66/tst/testallV11/poin.tst:5
# Input is:
RankHomologyPGroup(G,P,1000);
# Expected output:
251000
# But found:
The input must be a rational function follwed by a positive integer. 
Error, List Element: <list> must be a list (not the value 'fail')
########
```

With 4.14.0-beta1, I still get those two, but also a few others:

```
testing: /var/lib/portage/tmp/portage/dev-gap/hap-1.66/temp/pkg/hap-1.66/tst/testall3/3.9.3.tst
   ? `intvec`(GAP_intvec) = `bigintvec` is not supported
   ? expected `intvec` = `intvec`
   ? error occurred in or before STDIN line 3963: `, 1, intvec(1,1,1,2,2,3,4) );`
   ? expected intvec-expression. type 'help intvec;'
########> Diff in /var/lib/portage/tmp/portage/dev-gap/hap-1.66/temp/pkg/hap-1.66/tst/testall3/3.9.3.tst:4
# Input is:
P:=HilbertPoincareSeries(A);;
# Expected output:
# But found:
Error, no method found! For debugging hints type ?Recovery from NoMethodFound
Error, no 1st choice method found for `Remove' on 2 arguments
########
########> Diff in /var/lib/portage/tmp/portage/dev-gap/hap-1.66/temp/pkg/hap-1.66/tst/testall3/3.9.3.tst:5
# Input is:
RankHomologyPGroup(G,P,1001);
# Expected output:
251503
# But found:
The input must be a rational function follwed by a positive integer. 
Error, List Element: <list> must be a list (not the value 'fail')
...
testing: /var/lib/portage/tmp/portage/dev-gap/hap-1.66/temp/pkg/hap-1.66/tst/testallV11/cohodata.tst
   ? `intvec`(GAP_intvec) = `bigintvec` is not supported
   ? expected `intvec` = `intvec`
   ? error occurred in or before STDIN line 3981: `, 1, intvec(1,1,4) );`
   ? expected intvec-expression. type 'help intvec;'
########> Diff in /var/lib/portage/tmp/portage/dev-gap/hap-1.66/temp/pkg/hap-1.66/tst/testallV11/cohodata.tst:2
# Input is:
CohomologicalData(SmallGroup(8,4),20);
# Expected output:
Integer argument is large enough to ensure completeness of cohomology ring presentation.

Group order: 8
Group number: 4
Group description: Q8

Cohomology generators
Degree 1: a, b
Degree 4: c

Cohomology relations
1: a^2+a*b+b^2
2: b^3

Poincare series
(x^2+x+1)/(-x^3+x^2-x+1)

Steenrod squares
Sq^1(c)=0
Sq^2(c)=0

<algebra of dimension 15 over GF(2)>
# But found:
Integer argument is large enough to ensure completeness of cohomology ring presentation.

Error, no method found! For debugging hints type ?Recovery from NoMethodFound
Error, no 1st choice method found for `Remove' on 2 arguments

...

-----------------------------------
total    411750 ms (58485 ms GC) and 13.3GB allocated
              5 failures in 3 of 89 files

#I  Errors detected while testing

fgets() failed with errno 5
I/O error
```

**Reply by fingolfin** [reply]
The `hap` failure you describe suggest an issue with `Singular` (the messages involving `intvec` etc. look like they come from Singular). Perhaps you are using a newer Singular version than we use in CI and it once again has a breaking change. Not too much we can do about that right now then.

**Comment by jamesjer** [comment]
Yes, Singular 4.4 changed the return type of "hilb" from intvec to bigintvec.  Patching `lib/HapPrime/gradedalgebra.gi` to expect the bigintvec return type still leaves a failing test, though.  In `tst/testall1/1.3.2.txt`, the call to `BettiNumber(M,0)` on line 9 is expected to return 20, but returns 80 with Singular 4.4.

**Reply by fingolfin** [reply]
Singular unfortunately keeps making patch level releases that are ABI breaking. I've been unsuccessful in trying to stop that from happening but without success (despite sitting in the offices next to person doing this. Ah well).

Unfortunately the `singular` GAP package does not currently "know" about `bigintvec` (and support for Singular type `bigint` seems incomplete, too). I've filed https://github.com/gap-packages/singular/issues/15 but I am not sure anyone will pick it up. 

Once that is in place, hap could be changed to check the Singular version number and depending on which it is, could pass `"intvec"` or `"bigintvec"` as expected Singular return type.

This won't happen for GAP 4.14.0 and I am not sure when it might happen afterwards. But I also don't consider this is a blocker for GAP 4.14.0.

**Reply by orlitzky** [reply]
Ok, thanks, it's not too important I guess since I was the first to notice it. I opened separate bugs for the GCC 15 issues so that I don't forget. `-Werror=incompatible-pointer-types` is actually in GCC 14 already but the change to C23 by default in GCC 15 is making `foo()` mean `foo(void)`, and it's the combination of the two causing headaches.
